Feat/product

yjn33 opened the pull request:
## 📌 PR 요약
- 이 PR에서 어떤 작업이 이루어졌는지 간단히 설명해주세요.

- 상품정보수정 -> 말그대로 이미 등록된 product의 어트리뷰트를 수정한다
- 단건조회 -> id값으로 조회한다
- 삭제(중요) -> db상에서 데이터를 완전삭제가 아닌 soft delete로 구현
- 판매상태변경 -> soldOut boolean타입 값을 변경하는 기능 구현
- 상품에 이미지 추가 기능 구현


## 🔗 관련 이슈
- 관련된 이슈 번호를 연결해주세요.
    -해당 없음


## 🛠️ 변경 사항
- 주요 변경 사항을 bullet으로 작성해주세요.
    - /v1/products/{productId}
    - /v1/products/soldout/{productId}
    - /v1/products/delete/{productId}


## 📸 스크린샷 (선택)
UI 변경 사항이 있다면 스크린샷을 첨부해주세요.

상품 등록시 이미지가 정상적으로 등록되어지는것을 볼 수 있습니다.
![스크린샷 2025-03-27 16-31-23](https://github.com/user-attachments/assets/3b50510d-8333-4bcf-80cf-71204125b607)


상품 다건 단건 조회시 이미지가 조회 됩니다.
![스크린샷 2025-03-27 16-31-36](https://github.com/user-attachments/assets/a2b2ab34-47f5-4f15-8d3f-26ab77753cec)

![스크린샷 2025-03-27 16-31-45](https://github.com/user-attachments/assets/31f31aa2-7656-43a3-850d-e49fcd24c09c)


판매상태 변경시 soldOut 변수가 정상적으로 변경됩니다.
![스크린샷 2025-03-27 16-32-07](https://github.com/user-attachments/assets/b1359a6b-e96c-4f4f-98b5-b0c395ba84cc)


상품 정보 수정도 잘 됩니다.
![스크린샷 2025-03-27 16-32-13](https://github.com/user-attachments/assets/308a038e-3952-44a3-81a8-3bee59673f8e)


상품 소프트 딜리트도 잘됩니다. ( GetMapping -> deleteMapping으로 수정)
![스크린샷 2025-03-27 19-46-58](https://github.com/user-attachments/assets/fcbd31a2-6ab4-42e5-ae53-23673144c25b)


소프트 딜리트 되어진 상품은 deletedAt에 날짜가 적히고 조회가 안됩니다.
![스크린샷 2025-03-27 16-32-25](https://github.com/user-attachments/assets/7f1d72eb-17a2-4052-a6b8-02cb831c5ebd)

![스크린샷 2025-03-27 16-49-36](https://github.com/user-attachments/assets/c226867b-d3bf-443c-bc26-ce743eaeabc0)


## ⚠️ 주의 사항 (선택)
리뷰어가 알아야 할 중요한 사항이나 주의할 점을 작성해주세요.

http://localhost:8080/api/v1/products
상품등록 json
```
{
    "description" : "description 상품 상세 설명",
    "name" : "name 상품이름",
    "image" : "1",
    "stock" : "5",
    "price" : "5000"
}
```

http://localhost:8080/api/v1/products/soldout/1
상품 판매상태 변경 json
```
{
    "soldOut" : true
}
```

http://localhost:8080/api/v1/products/1
상품 정보수정 json
```
{
    "description" : "상세내용수정",
    "name" : "이름도수정",
    "image" : "1",
    "stock" : "2",
    "price" : "1111"
}
```

http://localhost:8080/api/v1/products/delete/1
상품 소프트 딜리트


## ✅ 체크리스트
PR 작성자가 확인해야 할 항목입니다:
- [x] 코드가 정상적으로 동작하는지 확인했습니다.
- [x] 관련 문서를 업데이트했습니다.
- [x] 코드 리뷰어를 등록했습니다.
- [x] Labels를 등록했습니다.

## :earth_asia: 그외 예상되는 버그

치명적인 버그는 아니지만 자잘한 버그들이 있습니다.

다건조회에서 소프트딜리트 레코드 조회안됨 -> 정상
소프트딜리트 -> 한번 삭제하고 다시 하면 계속 딜리트 응답됨 -> 비정상
단건조회에서 소프트딜리트해도 조회됨 -> 비정상 

request, response dto를 각자 구현해야 응답이 깔끔하게 나올 것 같습니다.
자잘한 버그들은 Setter 어노테이션과 함께 리팩토링 하겠습니다.

yjn33 committed:
Perf: Product (상품) 객체 소프트 딜리트 기준 설정

ProductService.java, ProductRepository.java -> 소프트 딜리트 기능구현 대비를 위해 findByDeletedAtNull 쿼리메소드 적용

Product.java -> Product(상품) 등록시 자동으로 deletedAt을 null 로 설정하기 위해 @ColumnDefault("null") 적용

yjn33 committed:
Feat: Product (상품) 단건 조회 기능 구현

yjn33 committed:
Feat: Product (상품) soft delete 기능 구현

yjn33 committed:
Feat: Product (상품) soldOut (판매상태) 변경 기능

soldOut boolean을 변경하여 현재 상품의 판매 상태를 설정

yjn33 committed:
Refactor: Product (상품) soldOut (판매상태) 어노테이션 변경

@PostMapping -> @PatchMapping 으로 변경

yjn33 committed:
Feat: Product (상품) 정보 수정 기능 구현, 타인이 등록한 상품 판매상태 변경 방지 로직 추가

editProduct -> 등록된 상품을 조회하여 정보를 수정하는 기능

setSoldOut -> AuthUser 를 통해 타인이 등록한 상품의 판매상태 변경 방지 로직 구현

yjn33 committed:
Merge branch 'dev' into feat/product

yjn33 committed:
Feat: 상품 등록, 정보수정시 Long image 를 통해 상품에 이미지 추가기능 구현

yjn33 committed:
Merge branch 'dev' into feat/product

RyuSungHyeon990604 commented on src/main/java/com/example/ddakdaegi/domain/product/controller/ProductController.java:
```diff
@@ -47,16 +51,67 @@ public Response<ProductResponse> saveProduct( //public ResponseEntity<ProductRes
+	/*
+		상품 삭제 소프트딜리트 메서드
+	*/
+	@GetMapping("/v1/products/delete/{productId}")
```
Get 보다는 delete 또는 post가 좋을거같습니다

RyuSungHyeon990604 reviewed (approved):
확인했습니다

yjn33 committed:
Refactor: pull request Feat/product #29 사항에 따른 코드 수정

ProductController.java -> 상품 소프트딜리트 매핑 어노테이션 수정 @GetMapping -> @DeleteMapping

ProductService.java -> imageRepository 조회 코드 리팩토링
